# Data Cleaning

In this notebook, we will be looking and cleaning the data files that is found in the Kaggle competition.

1. `weather.csv` ([link](#Clean-weather.csv))
2. `train.csv` and `test.csv` ([link](#Clean-train.csv-and-test.csv))
3. `spray.csv` ([link](#Clean-spray.csv))

---

## Clean weather.csv

In [1]:
# For Calculation and Data Manipulation
import numpy as np
import pandas as pd

In [2]:
# both file will be located in the same folder
original_weather_filename = '../assets/weather.csv'
clean_weather_filename = '../assets/cleaned_weather.csv'

In [3]:
# read in data
weather = pd.read_csv(original_weather_filename)

In [4]:
# see data shape and first 5 rows
print(f'shape: {weather.shape}')
weather.head()

shape: (2944, 22)


Station        Date  Tmax  Tmin Tavg Depart  DewPoint WetBulb Heat Cool  \
0        1  2007-05-01    83    50   67     14        51      56    0    2   
1        2  2007-05-01    84    52   68      M        51      57    0    3   
2        1  2007-05-02    59    42   51     -3        42      47   14    0   
3        2  2007-05-02    60    43   52      M        42      47   13    0   
4        1  2007-05-03    66    46   56      2        40      48    9    0   

   ... CodeSum Depth Water1 SnowFall PrecipTotal StnPressure SeaLevel  \
0  ...             0      M      0.0        0.00       29.10    29.82   
1  ...             M      M        M        0.00       29.18    29.82   
2  ...      BR     0      M      0.0        0.00       29.38    30.09   
3  ...   BR HZ     M      M        M        0.00       29.44    30.08   
4  ...             0      M      0.0        0.00       29.39    30.12   

  ResultSpeed ResultDir  AvgSpeed  
0         1.7        27       9.2  
1         2.7        25       9.6  
2        13.0         4      13.4  
3        13.3         2      13.4  
4        11.7         7      11.9  

[5 rows x 22 columns]

In [5]:
# see data info and if there is any null values
weather.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2944 entries, 0 to 2943
Data columns (total 22 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Station      2944 non-null   int64  
 1   Date         2944 non-null   object 
 2   Tmax         2944 non-null   int64  
 3   Tmin         2944 non-null   int64  
 4   Tavg         2944 non-null   object 
 5   Depart       2944 non-null   object 
 6   DewPoint     2944 non-null   int64  
 7   WetBulb      2944 non-null   object 
 8   Heat         2944 non-null   object 
 9   Cool         2944 non-null   object 
 10  Sunrise      2944 non-null   object 
 11  Sunset       2944 non-null   object 
 12  CodeSum      2944 non-null   object 
 13  Depth        2944 non-null   object 
 14  Water1       2944 non-null   object 
 15  SnowFall     2944 non-null   object 
 16  PrecipTotal  2944 non-null   object 
 17  StnPressure  2944 non-null   object 
 18  SeaLevel     2944 non-null   object 
 19  Result

Let's change the DType for `Date`.

In [6]:
weather['Date'] = pd.to_datetime(weather['Date'])

# see data info
weather.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2944 entries, 0 to 2943
Data columns (total 22 columns):
 #   Column       Non-Null Count  Dtype         
---  ------       --------------  -----         
 0   Station      2944 non-null   int64         
 1   Date         2944 non-null   datetime64[ns]
 2   Tmax         2944 non-null   int64         
 3   Tmin         2944 non-null   int64         
 4   Tavg         2944 non-null   object        
 5   Depart       2944 non-null   object        
 6   DewPoint     2944 non-null   int64         
 7   WetBulb      2944 non-null   object        
 8   Heat         2944 non-null   object        
 9   Cool         2944 non-null   object        
 10  Sunrise      2944 non-null   object        
 11  Sunset       2944 non-null   object        
 12  CodeSum      2944 non-null   object        
 13  Depth        2944 non-null   object        
 14  Water1       2944 non-null   object        
 15  SnowFall     2944 non-null   object        
 16  Precip

For easier manipulation, we will create 3 columns to store the day, month and year of the date.

In [7]:
# create 3 columns to store the day, month and year
weather['day'] = weather['Date'].dt.day
weather['month'] = weather['Date'].dt.month
weather['year'] = weather['Date'].dt.year

# see data info
weather.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2944 entries, 0 to 2943
Data columns (total 25 columns):
 #   Column       Non-Null Count  Dtype         
---  ------       --------------  -----         
 0   Station      2944 non-null   int64         
 1   Date         2944 non-null   datetime64[ns]
 2   Tmax         2944 non-null   int64         
 3   Tmin         2944 non-null   int64         
 4   Tavg         2944 non-null   object        
 5   Depart       2944 non-null   object        
 6   DewPoint     2944 non-null   int64         
 7   WetBulb      2944 non-null   object        
 8   Heat         2944 non-null   object        
 9   Cool         2944 non-null   object        
 10  Sunrise      2944 non-null   object        
 11  Sunset       2944 non-null   object        
 12  CodeSum      2944 non-null   object        
 13  Depth        2944 non-null   object        
 14  Water1       2944 non-null   object        
 15  SnowFall     2944 non-null   object        
 16  Precip

We can see that there are no null values, however, reading the data dictionary, it indicates that if there is missing data, it will be indicated by `M`.

In [8]:
# show the column and the number of missing data
for col in weather.columns:
    if(weather.loc[weather[col]=='M',[col]].count())[0]>0:
        print(f"{col}: {(weather.loc[weather[col]=='M',[col]].count())[0]}")

Tavg: 11
Depart: 1472
WetBulb: 4
Heat: 11
Cool: 11
Depth: 1472
Water1: 2944
SnowFall: 1472
PrecipTotal: 2
StnPressure: 4
SeaLevel: 9
AvgSpeed: 3


In [9]:
for col in weather.columns:
    if(weather.loc[weather[col]=='-',[col]].count())[0]>0:
        print(f"{col}: {(weather.loc[weather[col]=='-',[col]].count())[0]}")

Sunrise: 1472
Sunset: 1472


In [10]:
for col in weather.columns:
    if(weather.loc[weather[col]=='  T',[col]].count())[0]>0:
        print(f"{col}: {(weather.loc[weather[col]=='  T',[col]].count())[0]}")

SnowFall: 12
PrecipTotal: 318


As indicated in the code above, half of the entries (1472 out of 2944) in `depart`, `sunrise` and `sunset` columns have missing values. For `sunrise` and `sunset` columns, after some desktop research and understand that this was because station 2 does not collect data for these columns. We have thus decided to impute these missing values with values from station 1.

For column `water1`, all of its entries are missing values `M`, hence we should drop this column. We have also decided to drop `codesum`, `snowfall` and `depth` columns as their entries are either 0 or missing.

There are `  T` values in `snowfall` and `precitotal`. Based on the data documentation, this means that there are trace precipitate for that entry. Hence, we decided to round these values to 0. For the missing values in `preciptotal`, we decided to impute it with median values. Likewise for `avgspeed`, `sealevel` and `stnpressure`.

In [11]:
# replace 'sunrise', 'sunset' and 'depart'
# replace value from station 2 with station 1
weather_backup = weather.copy()

i = 0
while i < weather.shape[0]:
    weather.iloc[i+1, 5] = weather.iloc[i, 5]
    weather.iloc[i+1, 10] = weather.iloc[i, 10]
    weather.iloc[i+1, 11] = weather.iloc[i, 11]
    i+=2

In [12]:
weather.drop(columns = ['CodeSum','Water1','SnowFall','Depth'], inplace = True)

In [13]:
def impute_missing_tavg(row):
    if row['Tavg'] == 'M': 
        row['Tavg'] = (row['Tmax'] - row['Tmin']) * 0.5 + row['Tmin']
    return row

weather = weather.apply(impute_missing_tavg, axis = 1)
weather['Tavg'] = weather['Tavg'].astype('int64')

In [14]:
def impute_missing_wetbulb(row): 
    if row['WetBulb'] == 'M':
        row['WetBulb'] = row['Tavg']-((row['Tavg']-row['DewPoint'])/3)
    return row

weather = weather.apply(impute_missing_wetbulb, axis = 1)

In [15]:
def impute_missing_rest(row): 
    if row['Heat'] == 'M':
        if row['Tavg'] >= 65: 
            row['Heat'] = 0
            row['Cool'] = row['Tavg'] - 65
        else: 
            row['Heat'] = 65 - row['Tavg']
            row['Cool'] = 0

    if row['PrecipTotal'] == '  T':
        row['PrecipTotal'] = 0
    if row['PrecipTotal'] == 'M':
        row['PrecipTotal'] = weather.loc[(weather['PrecipTotal']!='M')&(weather['PrecipTotal']!='  T'),['PrecipTotal']].median()[0]       
    if row['StnPressure'] == 'M':
        row['StnPressure'] = weather[weather['StnPressure']!='M']['StnPressure'].median()
    if row['SeaLevel'] == 'M':
        row['SeaLevel'] = weather[weather['SeaLevel']!='M']['SeaLevel'].median()
    if row['AvgSpeed'] == 'M':
        row['AvgSpeed'] = weather[weather['AvgSpeed']!='M']['AvgSpeed'].median()    
    return row

weather = weather.apply(impute_missing_rest, axis = 1)

We check if there is any more missing data after cleaning.

In [16]:
# show the column and the number of missing data
for col in weather.columns:
    if(weather.loc[weather[col]=='M',[col]].count())[0]>0:
        print(f"{col}: {(weather.loc[weather[col]=='M',[col]].count())[0]}")

In [17]:
for col in weather.columns:
    if(weather.loc[weather[col]=='-',[col]].count())[0]>0:
        print(f"{col}: {(weather.loc[weather[col]=='-',[col]].count())[0]}")

In [18]:
for col in weather.columns:
    if(weather.loc[weather[col]=='  T',[col]].count())[0]>0:
        print(f"{col}: {(weather.loc[weather[col]=='  T',[col]].count())[0]}")

In [19]:
weather.head()

Station       Date  Tmax  Tmin  Tavg Depart  DewPoint WetBulb Heat Cool  \
0        1 2007-05-01    83    50    67     14        51      56    0    2   
1        2 2007-05-01    84    52    68     14        51      57    0    3   
2        1 2007-05-02    59    42    51     -3        42      47   14    0   
3        2 2007-05-02    60    43    52     -3        42      47   13    0   
4        1 2007-05-03    66    46    56      2        40      48    9    0   

   ... Sunset PrecipTotal StnPressure SeaLevel ResultSpeed  ResultDir  \
0  ...   1849        0.00       29.10    29.82         1.7         27   
1  ...   1849        0.00       29.18    29.82         2.7         25   
2  ...   1850        0.00       29.38    30.09        13.0          4   
3  ...   1850        0.00       29.44    30.08        13.3          2   
4  ...   1851        0.00       29.39    30.12        11.7          7   

   AvgSpeed day  month  year  
0       9.2   1      5  2007  
1       9.6   1      5  2007  
2      13.4   2      5  2007  
3      13.4   2      5  2007  
4      11.9   3      5  2007  

[5 rows x 21 columns]

With no more missing values, we will save the processed dataset.

In [20]:
weather.to_csv(clean_weather_filename)

In [65]:
# save weather file to pkl format
weather.to_pickle('../assets/clean_weather.pk1')

---

## Clean train.csv and test.csv

In [21]:
# both file will be located in the same folder
original_train_filename = '../assets/train.csv'
clean_train_filename = '../assets/cleaned_train.csv'

original_test_filename = '../assets/test.csv'
clean_test_filename = '../assets/cleaned_test.csv'

In [22]:
# read in data
train = pd.read_csv(original_train_filename)
test = pd.read_csv(original_test_filename)

In [23]:
# see data shape and first 5 rows
print(f'Train dataframe shape: {train.shape}')
train.head()

Train dataframe shape: (10506, 12)


Date                                            Address  \
0  2007-05-29  4100 North Oak Park Avenue, Chicago, IL 60634,...   
1  2007-05-29  4100 North Oak Park Avenue, Chicago, IL 60634,...   
2  2007-05-29  6200 North Mandell Avenue, Chicago, IL 60646, USA   
3  2007-05-29    7900 West Foster Avenue, Chicago, IL 60656, USA   
4  2007-05-29    7900 West Foster Avenue, Chicago, IL 60656, USA   

                  Species  Block           Street  Trap  \
0  CULEX PIPIENS/RESTUANS     41   N OAK PARK AVE  T002   
1          CULEX RESTUANS     41   N OAK PARK AVE  T002   
2          CULEX RESTUANS     62    N MANDELL AVE  T007   
3  CULEX PIPIENS/RESTUANS     79     W FOSTER AVE  T015   
4          CULEX RESTUANS     79     W FOSTER AVE  T015   

              AddressNumberAndStreet   Latitude  Longitude  AddressAccuracy  \
0  4100  N OAK PARK AVE, Chicago, IL  41.954690 -87.800991                9   
1  4100  N OAK PARK AVE, Chicago, IL  41.954690 -87.800991                9   
2   6200  N MANDELL AVE, Chicago, IL  41.994991 -87.769279                9   
3    7900  W FOSTER AVE, Chicago, IL  41.974089 -87.824812                8   
4    7900  W FOSTER AVE, Chicago, IL  41.974089 -87.824812                8   

   NumMosquitos  WnvPresent  
0             1           0  
1             1           0  
2             1           0  
3             1           0  
4             4           0

In [24]:
# see data shape and first 5 rows
print(f'Test dataframe shape: {test.shape}')
test.head()

Test dataframe shape: (116293, 11)


Id        Date                                            Address  \
0   1  2008-06-11  4100 North Oak Park Avenue, Chicago, IL 60634,...   
1   2  2008-06-11  4100 North Oak Park Avenue, Chicago, IL 60634,...   
2   3  2008-06-11  4100 North Oak Park Avenue, Chicago, IL 60634,...   
3   4  2008-06-11  4100 North Oak Park Avenue, Chicago, IL 60634,...   
4   5  2008-06-11  4100 North Oak Park Avenue, Chicago, IL 60634,...   

                  Species  Block           Street  Trap  \
0  CULEX PIPIENS/RESTUANS     41   N OAK PARK AVE  T002   
1          CULEX RESTUANS     41   N OAK PARK AVE  T002   
2           CULEX PIPIENS     41   N OAK PARK AVE  T002   
3        CULEX SALINARIUS     41   N OAK PARK AVE  T002   
4         CULEX TERRITANS     41   N OAK PARK AVE  T002   

              AddressNumberAndStreet  Latitude  Longitude  AddressAccuracy  
0  4100  N OAK PARK AVE, Chicago, IL  41.95469 -87.800991                9  
1  4100  N OAK PARK AVE, Chicago, IL  41.95469 -87.800991                9  
2  4100  N OAK PARK AVE, Chicago, IL  41.95469 -87.800991                9  
3  4100  N OAK PARK AVE, Chicago, IL  41.95469 -87.800991                9  
4  4100  N OAK PARK AVE, Chicago, IL  41.95469 -87.800991                9

In [25]:
# see data info and if there is any null values
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10506 entries, 0 to 10505
Data columns (total 12 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Date                    10506 non-null  object 
 1   Address                 10506 non-null  object 
 2   Species                 10506 non-null  object 
 3   Block                   10506 non-null  int64  
 4   Street                  10506 non-null  object 
 5   Trap                    10506 non-null  object 
 6   AddressNumberAndStreet  10506 non-null  object 
 7   Latitude                10506 non-null  float64
 8   Longitude               10506 non-null  float64
 9   AddressAccuracy         10506 non-null  int64  
 10  NumMosquitos            10506 non-null  int64  
 11  WnvPresent              10506 non-null  int64  
dtypes: float64(2), int64(4), object(6)
memory usage: 985.1+ KB


In [26]:
# see data info and if there is any null values
test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 116293 entries, 0 to 116292
Data columns (total 11 columns):
 #   Column                  Non-Null Count   Dtype  
---  ------                  --------------   -----  
 0   Id                      116293 non-null  int64  
 1   Date                    116293 non-null  object 
 2   Address                 116293 non-null  object 
 3   Species                 116293 non-null  object 
 4   Block                   116293 non-null  int64  
 5   Street                  116293 non-null  object 
 6   Trap                    116293 non-null  object 
 7   AddressNumberAndStreet  116293 non-null  object 
 8   Latitude                116293 non-null  float64
 9   Longitude               116293 non-null  float64
 10  AddressAccuracy         116293 non-null  int64  
dtypes: float64(2), int64(3), object(6)
memory usage: 9.8+ MB


Let's change the DType for `Date`.

In [27]:
train['Date'] = pd.to_datetime(train['Date'])
test['Date'] = pd.to_datetime(test['Date'])

In [28]:
# see data info and if there is any null values
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10506 entries, 0 to 10505
Data columns (total 12 columns):
 #   Column                  Non-Null Count  Dtype         
---  ------                  --------------  -----         
 0   Date                    10506 non-null  datetime64[ns]
 1   Address                 10506 non-null  object        
 2   Species                 10506 non-null  object        
 3   Block                   10506 non-null  int64         
 4   Street                  10506 non-null  object        
 5   Trap                    10506 non-null  object        
 6   AddressNumberAndStreet  10506 non-null  object        
 7   Latitude                10506 non-null  float64       
 8   Longitude               10506 non-null  float64       
 9   AddressAccuracy         10506 non-null  int64         
 10  NumMosquitos            10506 non-null  int64         
 11  WnvPresent              10506 non-null  int64         
dtypes: datetime64[ns](1), float64(2), int64(4), ob

In [29]:
# see data info and if there is any null values
test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 116293 entries, 0 to 116292
Data columns (total 11 columns):
 #   Column                  Non-Null Count   Dtype         
---  ------                  --------------   -----         
 0   Id                      116293 non-null  int64         
 1   Date                    116293 non-null  datetime64[ns]
 2   Address                 116293 non-null  object        
 3   Species                 116293 non-null  object        
 4   Block                   116293 non-null  int64         
 5   Street                  116293 non-null  object        
 6   Trap                    116293 non-null  object        
 7   AddressNumberAndStreet  116293 non-null  object        
 8   Latitude                116293 non-null  float64       
 9   Longitude               116293 non-null  float64       
 10  AddressAccuracy         116293 non-null  int64         
dtypes: datetime64[ns](1), float64(2), int64(3), object(5)
memory usage: 9.8+ MB


For `train.csv` and `test.csv`, the test results were organized in such a way that when the number of mosquitos exceed 50, they are split into another record (another row in the dataset), such that the number of mosquitos are capped at 50. 

We will merge the duplicated rows found in `train` which are read from `train.csv`.
As we will be trying to predict `wnvpresent` in `test.csv`, the values of `NumMosquitos` and `WnvPresent` are missing from the dataframe `test`, hence there is no need for the merging of the duplicated rows. 

In [30]:
# create new df for the merging and not affect original df
train_2 = train.copy()

# helps to ensure that the duplicated rows are one after the other
train_2.sort_values(by='Date')

# column to identify duplicated rows
cols = ['Date', 'Address', 'Species', 'Block', 'Street', 
                           'Trap', 'AddressNumberAndStreet', 'Latitude', 
                           'Longitude', 'AddressAccuracy']

# duplicated rows
# train_2[train_2.duplicated(cols, keep=False)]
# duplicated rows index
#[index for index, row in train_2.duplicated(cols, keep=False).iteritems() if row]

# train_2.groupby(by=cols)[['NumMosquitos', 'WnvPresent']]
train_2 = train_2.groupby(by=cols).agg({'NumMosquitos':np.sum, 'WnvPresent': np.sum})
train_2 = train_2.reset_index()

# see data shape and first 5 rows
print(f'Duplicated rows merged dataframe shape: {train_2.shape}')
train_2.head()

Duplicated rows merged dataframe shape: (8475, 12)


Date                                           Address  \
0 2007-05-29       1100 Roosevelt Road, Chicago, IL 60608, USA   
1 2007-05-29       1100 Roosevelt Road, Chicago, IL 60608, USA   
2 2007-05-29  1100 South Peoria Street, Chicago, IL 60608, USA   
3 2007-05-29  1100 West Chicago Avenue, Chicago, IL 60642, USA   
4 2007-05-29    1500 North Long Avenue, Chicago, IL 60651, USA   

                  Species  Block        Street  Trap  \
0  CULEX PIPIENS/RESTUANS     11   W ROOSEVELT  T048   
1          CULEX RESTUANS     11   W ROOSEVELT  T048   
2          CULEX RESTUANS     11   S PEORIA ST  T091   
3          CULEX RESTUANS     11     W CHICAGO  T049   
4          CULEX RESTUANS     15    N LONG AVE  T153   

           AddressNumberAndStreet   Latitude  Longitude  AddressAccuracy  \
0  1100  W ROOSEVELT, Chicago, IL  41.867108 -87.654224                8   
1  1100  W ROOSEVELT, Chicago, IL  41.867108 -87.654224                8   
2  1100  S PEORIA ST, Chicago, IL  41.862292 -87.648860                8   
3    1100  W CHICAGO, Chicago, IL  41.896282 -87.655232                8   
4   1500  N LONG AVE, Chicago, IL  41.907645 -87.760886                8   

   NumMosquitos  WnvPresent  
0             1           0  
1             2           0  
2             1           0  
3             1           0  
4             1           0

In [31]:
# see data info and if there is any null values
train_2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8475 entries, 0 to 8474
Data columns (total 12 columns):
 #   Column                  Non-Null Count  Dtype         
---  ------                  --------------  -----         
 0   Date                    8475 non-null   datetime64[ns]
 1   Address                 8475 non-null   object        
 2   Species                 8475 non-null   object        
 3   Block                   8475 non-null   int64         
 4   Street                  8475 non-null   object        
 5   Trap                    8475 non-null   object        
 6   AddressNumberAndStreet  8475 non-null   object        
 7   Latitude                8475 non-null   float64       
 8   Longitude               8475 non-null   float64       
 9   AddressAccuracy         8475 non-null   int64         
 10  NumMosquitos            8475 non-null   int64         
 11  WnvPresent              8475 non-null   int64         
dtypes: datetime64[ns](1), float64(2), int64(4), obje

In [32]:
# confirm that totals remains the same after merging
print(f"Before merging, NumMosquitos: {train['NumMosquitos'].sum()}")
print(f"Before merging, NumMosquitos: {train_2['NumMosquitos'].sum()}")
print()
print(f"Before merging, WnvPresent: {train['WnvPresent'].sum()}")
print(f"After merging, WnvPresent: {train_2['WnvPresent'].sum()}")

Before merging, NumMosquitos: 135039
Before merging, NumMosquitos: 135039

Before merging, WnvPresent: 551
After merging, WnvPresent: 551


In [33]:
# check unique values for WnvPresent
print(f"Unique values in train_2 for 'WnvPresent': {train_2['WnvPresent'].unique()}")

Unique values in train_2 for 'WnvPresent': [ 0  1  3  2 10  7  6  4  9  8  5]


Recall that `WnvPresent` should only be 0 or 1, we will use a list comprehension to replace the values more than 1. 

In [34]:
train_2['WnvPresent'] = [0 if value == 0 else 1 for value in train_2['WnvPresent']]
train_2['WnvPresent'].unique()

array([0, 1], dtype=int64)

For easier manipulation, we will create 3 columns to store the day, month and year of the date.

In [35]:
# create 3 columns to store the day, month and year

# for train
train_2['day'] = train_2['Date'].dt.day
train_2['month'] = train_2['Date'].dt.month
train_2['year'] = train_2['Date'].dt.year

# for test
test['day'] = test['Date'].dt.day
test['month'] = test['Date'].dt.month
test['year'] = test['Date'].dt.year

In [36]:
# see data shape and first 5 rows
print(f'Train dataframe shape: {train_2.shape}')
train_2.head()

Train dataframe shape: (8475, 15)


Date                                           Address  \
0 2007-05-29       1100 Roosevelt Road, Chicago, IL 60608, USA   
1 2007-05-29       1100 Roosevelt Road, Chicago, IL 60608, USA   
2 2007-05-29  1100 South Peoria Street, Chicago, IL 60608, USA   
3 2007-05-29  1100 West Chicago Avenue, Chicago, IL 60642, USA   
4 2007-05-29    1500 North Long Avenue, Chicago, IL 60651, USA   

                  Species  Block        Street  Trap  \
0  CULEX PIPIENS/RESTUANS     11   W ROOSEVELT  T048   
1          CULEX RESTUANS     11   W ROOSEVELT  T048   
2          CULEX RESTUANS     11   S PEORIA ST  T091   
3          CULEX RESTUANS     11     W CHICAGO  T049   
4          CULEX RESTUANS     15    N LONG AVE  T153   

           AddressNumberAndStreet   Latitude  Longitude  AddressAccuracy  \
0  1100  W ROOSEVELT, Chicago, IL  41.867108 -87.654224                8   
1  1100  W ROOSEVELT, Chicago, IL  41.867108 -87.654224                8   
2  1100  S PEORIA ST, Chicago, IL  41.862292 -87.648860                8   
3    1100  W CHICAGO, Chicago, IL  41.896282 -87.655232                8   
4   1500  N LONG AVE, Chicago, IL  41.907645 -87.760886                8   

   NumMosquitos  WnvPresent  day  month  year  
0             1           0   29      5  2007  
1             2           0   29      5  2007  
2             1           0   29      5  2007  
3             1           0   29      5  2007  
4             1           0   29      5  2007

In [37]:
# see data shape and first 5 rows
print(f'Test dataframe shape: {test.shape}')
test.head()

Test dataframe shape: (116293, 14)


Id       Date                                            Address  \
0   1 2008-06-11  4100 North Oak Park Avenue, Chicago, IL 60634,...   
1   2 2008-06-11  4100 North Oak Park Avenue, Chicago, IL 60634,...   
2   3 2008-06-11  4100 North Oak Park Avenue, Chicago, IL 60634,...   
3   4 2008-06-11  4100 North Oak Park Avenue, Chicago, IL 60634,...   
4   5 2008-06-11  4100 North Oak Park Avenue, Chicago, IL 60634,...   

                  Species  Block           Street  Trap  \
0  CULEX PIPIENS/RESTUANS     41   N OAK PARK AVE  T002   
1          CULEX RESTUANS     41   N OAK PARK AVE  T002   
2           CULEX PIPIENS     41   N OAK PARK AVE  T002   
3        CULEX SALINARIUS     41   N OAK PARK AVE  T002   
4         CULEX TERRITANS     41   N OAK PARK AVE  T002   

              AddressNumberAndStreet  Latitude  Longitude  AddressAccuracy  \
0  4100  N OAK PARK AVE, Chicago, IL  41.95469 -87.800991                9   
1  4100  N OAK PARK AVE, Chicago, IL  41.95469 -87.800991                9   
2  4100  N OAK PARK AVE, Chicago, IL  41.95469 -87.800991                9   
3  4100  N OAK PARK AVE, Chicago, IL  41.95469 -87.800991                9   
4  4100  N OAK PARK AVE, Chicago, IL  41.95469 -87.800991                9   

   day  month  year  
0   11      6  2008  
1   11      6  2008  
2   11      6  2008  
3   11      6  2008  
4   11      6  2008

In [38]:
# see data info and if there is any null values
train_2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8475 entries, 0 to 8474
Data columns (total 15 columns):
 #   Column                  Non-Null Count  Dtype         
---  ------                  --------------  -----         
 0   Date                    8475 non-null   datetime64[ns]
 1   Address                 8475 non-null   object        
 2   Species                 8475 non-null   object        
 3   Block                   8475 non-null   int64         
 4   Street                  8475 non-null   object        
 5   Trap                    8475 non-null   object        
 6   AddressNumberAndStreet  8475 non-null   object        
 7   Latitude                8475 non-null   float64       
 8   Longitude               8475 non-null   float64       
 9   AddressAccuracy         8475 non-null   int64         
 10  NumMosquitos            8475 non-null   int64         
 11  WnvPresent              8475 non-null   int64         
 12  day                     8475 non-null   int64   

In [39]:
# see data info and if there is any null values
test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 116293 entries, 0 to 116292
Data columns (total 14 columns):
 #   Column                  Non-Null Count   Dtype         
---  ------                  --------------   -----         
 0   Id                      116293 non-null  int64         
 1   Date                    116293 non-null  datetime64[ns]
 2   Address                 116293 non-null  object        
 3   Species                 116293 non-null  object        
 4   Block                   116293 non-null  int64         
 5   Street                  116293 non-null  object        
 6   Trap                    116293 non-null  object        
 7   AddressNumberAndStreet  116293 non-null  object        
 8   Latitude                116293 non-null  float64       
 9   Longitude               116293 non-null  float64       
 10  AddressAccuracy         116293 non-null  int64         
 11  day                     116293 non-null  int64         
 12  month                   116293

Now that we have cleaned the two dataframe, we will combine the two together for any analysis purpose. 

In [40]:
# identify column name in test but not in train_2
print(f"Column in test but not in train: {[col for col in test.columns if col not in train_2.columns]}")

# identify column name in train_2 but not in test
print(f"Column in train but not in test: {[col for col in train_2.columns if col not in test.columns]}")

Column in test but not in train: ['Id']
Column in train but not in test: ['NumMosquitos', 'WnvPresent']


In [41]:
# create copy for concat
train_concat = train_2.copy()

# add in column identifier that identifies if the data belongs to train or test
train_concat['is_train'] = 1

# add in missing column value compared to test
train_concat['Id'] = -1

In [42]:
# create copy for concat
test_concat = test.copy()

# add in column identifier that identifies if the data belongs to train or test
test_concat['is_train'] = 0

# add in missing column value compared to test
test_concat['NumMosquitos'] = -1
test_concat['WnvPresent'] = -1

In [43]:
# see data shape and first 5 rows
print(f'Train for concat dataframe shape: {train_concat.shape}')
train_concat.head()

Train for concat dataframe shape: (8475, 17)


Date                                           Address  \
0 2007-05-29       1100 Roosevelt Road, Chicago, IL 60608, USA   
1 2007-05-29       1100 Roosevelt Road, Chicago, IL 60608, USA   
2 2007-05-29  1100 South Peoria Street, Chicago, IL 60608, USA   
3 2007-05-29  1100 West Chicago Avenue, Chicago, IL 60642, USA   
4 2007-05-29    1500 North Long Avenue, Chicago, IL 60651, USA   

                  Species  Block        Street  Trap  \
0  CULEX PIPIENS/RESTUANS     11   W ROOSEVELT  T048   
1          CULEX RESTUANS     11   W ROOSEVELT  T048   
2          CULEX RESTUANS     11   S PEORIA ST  T091   
3          CULEX RESTUANS     11     W CHICAGO  T049   
4          CULEX RESTUANS     15    N LONG AVE  T153   

           AddressNumberAndStreet   Latitude  Longitude  AddressAccuracy  \
0  1100  W ROOSEVELT, Chicago, IL  41.867108 -87.654224                8   
1  1100  W ROOSEVELT, Chicago, IL  41.867108 -87.654224                8   
2  1100  S PEORIA ST, Chicago, IL  41.862292 -87.648860                8   
3    1100  W CHICAGO, Chicago, IL  41.896282 -87.655232                8   
4   1500  N LONG AVE, Chicago, IL  41.907645 -87.760886                8   

   NumMosquitos  WnvPresent  day  month  year  is_train  Id  
0             1           0   29      5  2007         1  -1  
1             2           0   29      5  2007         1  -1  
2             1           0   29      5  2007         1  -1  
3             1           0   29      5  2007         1  -1  
4             1           0   29      5  2007         1  -1

In [44]:
# see data shape and first 5 rows
print(f'Test for concat dataframe shape: {test_concat.shape}')
test_concat.head()

Test for concat dataframe shape: (116293, 17)


Id       Date                                            Address  \
0   1 2008-06-11  4100 North Oak Park Avenue, Chicago, IL 60634,...   
1   2 2008-06-11  4100 North Oak Park Avenue, Chicago, IL 60634,...   
2   3 2008-06-11  4100 North Oak Park Avenue, Chicago, IL 60634,...   
3   4 2008-06-11  4100 North Oak Park Avenue, Chicago, IL 60634,...   
4   5 2008-06-11  4100 North Oak Park Avenue, Chicago, IL 60634,...   

                  Species  Block           Street  Trap  \
0  CULEX PIPIENS/RESTUANS     41   N OAK PARK AVE  T002   
1          CULEX RESTUANS     41   N OAK PARK AVE  T002   
2           CULEX PIPIENS     41   N OAK PARK AVE  T002   
3        CULEX SALINARIUS     41   N OAK PARK AVE  T002   
4         CULEX TERRITANS     41   N OAK PARK AVE  T002   

              AddressNumberAndStreet  Latitude  Longitude  AddressAccuracy  \
0  4100  N OAK PARK AVE, Chicago, IL  41.95469 -87.800991                9   
1  4100  N OAK PARK AVE, Chicago, IL  41.95469 -87.800991                9   
2  4100  N OAK PARK AVE, Chicago, IL  41.95469 -87.800991                9   
3  4100  N OAK PARK AVE, Chicago, IL  41.95469 -87.800991                9   
4  4100  N OAK PARK AVE, Chicago, IL  41.95469 -87.800991                9   

   day  month  year  is_train  NumMosquitos  WnvPresent  
0   11      6  2008         0            -1          -1  
1   11      6  2008         0            -1          -1  
2   11      6  2008         0            -1          -1  
3   11      6  2008         0            -1          -1  
4   11      6  2008         0            -1          -1

In [45]:
# see data info and if there is any null values
train_concat.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8475 entries, 0 to 8474
Data columns (total 17 columns):
 #   Column                  Non-Null Count  Dtype         
---  ------                  --------------  -----         
 0   Date                    8475 non-null   datetime64[ns]
 1   Address                 8475 non-null   object        
 2   Species                 8475 non-null   object        
 3   Block                   8475 non-null   int64         
 4   Street                  8475 non-null   object        
 5   Trap                    8475 non-null   object        
 6   AddressNumberAndStreet  8475 non-null   object        
 7   Latitude                8475 non-null   float64       
 8   Longitude               8475 non-null   float64       
 9   AddressAccuracy         8475 non-null   int64         
 10  NumMosquitos            8475 non-null   int64         
 11  WnvPresent              8475 non-null   int64         
 12  day                     8475 non-null   int64   

In [46]:
# see data info and if there is any null values
test_concat.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 116293 entries, 0 to 116292
Data columns (total 17 columns):
 #   Column                  Non-Null Count   Dtype         
---  ------                  --------------   -----         
 0   Id                      116293 non-null  int64         
 1   Date                    116293 non-null  datetime64[ns]
 2   Address                 116293 non-null  object        
 3   Species                 116293 non-null  object        
 4   Block                   116293 non-null  int64         
 5   Street                  116293 non-null  object        
 6   Trap                    116293 non-null  object        
 7   AddressNumberAndStreet  116293 non-null  object        
 8   Latitude                116293 non-null  float64       
 9   Longitude               116293 non-null  float64       
 10  AddressAccuracy         116293 non-null  int64         
 11  day                     116293 non-null  int64         
 12  month                   116293

In [47]:
# concat the two data
train_test = pd.concat([train_concat, test_concat], ignore_index=True)

# see data shape and first 5 rows
print(f'combined dataframe shape: {train_test.shape}')
train_test.head()

combined dataframe shape: (124768, 17)


Date                                           Address  \
0 2007-05-29       1100 Roosevelt Road, Chicago, IL 60608, USA   
1 2007-05-29       1100 Roosevelt Road, Chicago, IL 60608, USA   
2 2007-05-29  1100 South Peoria Street, Chicago, IL 60608, USA   
3 2007-05-29  1100 West Chicago Avenue, Chicago, IL 60642, USA   
4 2007-05-29    1500 North Long Avenue, Chicago, IL 60651, USA   

                  Species  Block        Street  Trap  \
0  CULEX PIPIENS/RESTUANS     11   W ROOSEVELT  T048   
1          CULEX RESTUANS     11   W ROOSEVELT  T048   
2          CULEX RESTUANS     11   S PEORIA ST  T091   
3          CULEX RESTUANS     11     W CHICAGO  T049   
4          CULEX RESTUANS     15    N LONG AVE  T153   

           AddressNumberAndStreet   Latitude  Longitude  AddressAccuracy  \
0  1100  W ROOSEVELT, Chicago, IL  41.867108 -87.654224                8   
1  1100  W ROOSEVELT, Chicago, IL  41.867108 -87.654224                8   
2  1100  S PEORIA ST, Chicago, IL  41.862292 -87.648860                8   
3    1100  W CHICAGO, Chicago, IL  41.896282 -87.655232                8   
4   1500  N LONG AVE, Chicago, IL  41.907645 -87.760886                8   

   NumMosquitos  WnvPresent  day  month  year  is_train  Id  
0             1           0   29      5  2007         1  -1  
1             2           0   29      5  2007         1  -1  
2             1           0   29      5  2007         1  -1  
3             1           0   29      5  2007         1  -1  
4             1           0   29      5  2007         1  -1

In [48]:
# see data info and if there is any null values
train_test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 124768 entries, 0 to 124767
Data columns (total 17 columns):
 #   Column                  Non-Null Count   Dtype         
---  ------                  --------------   -----         
 0   Date                    124768 non-null  datetime64[ns]
 1   Address                 124768 non-null  object        
 2   Species                 124768 non-null  object        
 3   Block                   124768 non-null  int64         
 4   Street                  124768 non-null  object        
 5   Trap                    124768 non-null  object        
 6   AddressNumberAndStreet  124768 non-null  object        
 7   Latitude                124768 non-null  float64       
 8   Longitude               124768 non-null  float64       
 9   AddressAccuracy         124768 non-null  int64         
 10  NumMosquitos            124768 non-null  int64         
 11  WnvPresent              124768 non-null  int64         
 12  day                     124768

### Export the cleaned data

In [50]:
# Export cleaned data files

train_2.to_csv(clean_train_filename, index=False)
test.to_csv(clean_test_filename, index=False)
train_test.to_csv("../assets/cleaned_train_test.csv", index=False)

---

## Clean spray.csv

In [53]:
# both file will be located in the same folder
original_filename = '../assets/spray.csv'
clean_filename = '../assets/cleaned_spray.csv'

In [54]:
# read in data
spray = pd.read_csv(original_filename)

# see data shape and first 5 rows
print(spray.shape)
spray.head()

(14835, 4)


Date        Time   Latitude  Longitude
0  2011-08-29  6:56:58 PM  42.391623 -88.089163
1  2011-08-29  6:57:08 PM  42.391348 -88.089163
2  2011-08-29  6:57:18 PM  42.391022 -88.089157
3  2011-08-29  6:57:28 PM  42.390637 -88.089158
4  2011-08-29  6:57:38 PM  42.390410 -88.088858

In [55]:
# see data info and if there is any null values
spray.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14835 entries, 0 to 14834
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   Date       14835 non-null  object 
 1   Time       14251 non-null  object 
 2   Latitude   14835 non-null  float64
 3   Longitude  14835 non-null  float64
dtypes: float64(2), object(2)
memory usage: 463.7+ KB


Let's change the date DType for `Date` and create a new column to store the time in 24 hours format

In [56]:
# convert date Dtype
spray['Date'] = pd.to_datetime(spray['Date'])

# create new column to store time in 24 hour format
spray['Time_24h'] = pd.to_datetime(spray['Time']).dt.time

In [57]:
# see data shape and first 5 rows
print(spray.shape)
spray.head()

(14835, 5)


Date        Time   Latitude  Longitude  Time_24h
0 2011-08-29  6:56:58 PM  42.391623 -88.089163  18:56:58
1 2011-08-29  6:57:08 PM  42.391348 -88.089163  18:57:08
2 2011-08-29  6:57:18 PM  42.391022 -88.089157  18:57:18
3 2011-08-29  6:57:28 PM  42.390637 -88.089158  18:57:28
4 2011-08-29  6:57:38 PM  42.390410 -88.088858  18:57:38

In [58]:
# see data info
spray.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14835 entries, 0 to 14834
Data columns (total 5 columns):
 #   Column     Non-Null Count  Dtype         
---  ------     --------------  -----         
 0   Date       14835 non-null  datetime64[ns]
 1   Time       14251 non-null  object        
 2   Latitude   14835 non-null  float64       
 3   Longitude  14835 non-null  float64       
 4   Time_24h   14251 non-null  object        
dtypes: datetime64[ns](1), float64(2), object(2)
memory usage: 579.6+ KB


For easier manipulation, we will create 3 columns to store the day, month and year of the date. 

In [59]:
# create 3 columns to store the day, month and year
spray['day'] = spray['Date'].dt.day
spray['month'] = spray['Date'].dt.month
spray['year'] = spray['Date'].dt.year

In [60]:
# see data shape and first 5 rows
print(spray.shape)
spray.head()

(14835, 8)


Date        Time   Latitude  Longitude  Time_24h  day  month  year
0 2011-08-29  6:56:58 PM  42.391623 -88.089163  18:56:58   29      8  2011
1 2011-08-29  6:57:08 PM  42.391348 -88.089163  18:57:08   29      8  2011
2 2011-08-29  6:57:18 PM  42.391022 -88.089157  18:57:18   29      8  2011
3 2011-08-29  6:57:28 PM  42.390637 -88.089158  18:57:28   29      8  2011
4 2011-08-29  6:57:38 PM  42.390410 -88.088858  18:57:38   29      8  2011

In [61]:
# see data info
spray.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14835 entries, 0 to 14834
Data columns (total 8 columns):
 #   Column     Non-Null Count  Dtype         
---  ------     --------------  -----         
 0   Date       14835 non-null  datetime64[ns]
 1   Time       14251 non-null  object        
 2   Latitude   14835 non-null  float64       
 3   Longitude  14835 non-null  float64       
 4   Time_24h   14251 non-null  object        
 5   day        14835 non-null  int64         
 6   month      14835 non-null  int64         
 7   year       14835 non-null  int64         
dtypes: datetime64[ns](1), float64(2), int64(3), object(2)
memory usage: 927.3+ KB


In [62]:
# rename the columns to just lower case
spray.rename(columns={column: column.lower() for column in spray.columns}, inplace=True)

# see data shape and first 5 rows
print(spray.shape)
spray.head()

(14835, 8)


date        time   latitude  longitude  time_24h  day  month  year
0 2011-08-29  6:56:58 PM  42.391623 -88.089163  18:56:58   29      8  2011
1 2011-08-29  6:57:08 PM  42.391348 -88.089163  18:57:08   29      8  2011
2 2011-08-29  6:57:18 PM  42.391022 -88.089157  18:57:18   29      8  2011
3 2011-08-29  6:57:28 PM  42.390637 -88.089158  18:57:28   29      8  2011
4 2011-08-29  6:57:38 PM  42.390410 -88.088858  18:57:38   29      8  2011

In [63]:
# see data info
spray.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14835 entries, 0 to 14834
Data columns (total 8 columns):
 #   Column     Non-Null Count  Dtype         
---  ------     --------------  -----         
 0   date       14835 non-null  datetime64[ns]
 1   time       14251 non-null  object        
 2   latitude   14835 non-null  float64       
 3   longitude  14835 non-null  float64       
 4   time_24h   14251 non-null  object        
 5   day        14835 non-null  int64         
 6   month      14835 non-null  int64         
 7   year       14835 non-null  int64         
dtypes: datetime64[ns](1), float64(2), int64(3), object(2)
memory usage: 927.3+ KB


We will **not be removing the null data** from the data as we are not clear if time will be utilised in the analysis. As any point in time, we can remove these data by dropping the null data or dropping the column. 

```python
# to run the below code to drop null data
spray.dropna(inplace=True)

# to run below code to drop column
df.drop(['time', 'time_24h'], axis = 1, inplace = True)

```

In [64]:
# export cleaned data file
spray.to_csv(clean_filename, index=False)